### PR TITLE
[libc][stdfix] Implement `countlsfx` functions

### LINKED ISSUE #113357: Implement fixed point countls functions in llvm-libc
Some fixed point functions from ISO 18037 are implemented in llvm-libc, but not all of them are implemented. The various countls`fx` functions should also be added. Copying from the extension:

```
4.1.7.3 The fixed-point bit countls functions

The bit count functions countlsfx, where fx stands for one of hr, r, lr, hk, k, lk, uhr, ur, ulr, uhk, uk or ulk, take one fixed-point type argument (corresponding to fx); the result type is int.
The integer return value of the above functions is defined as follows:
- if the value of the fixed-point argument is non-zero, the return value is the largest integer k for
which the expression a<<k does not overflow;
- if the value of the fixed-point argument is zero, an integer value is returned that is at least as large as N, where N is the total number of value bits of the fixed-point type of the argument.
```

This effectively returns the number of leading sign bits. These can first be implemented as `__builtin_*` functions in clang then llvm-libc can provide the `countlsfx` wrappers for each of the builtin functions.

### PR DESCRIPTION
**Pull request by duncpro**
Closes #113357

### CODE REVIEW

File: libc/src/stdfix/CMakeLists.txt
Comment (lntue) [inline libc/src/stdfix/CMakeLists.txt:5]: nit: you can let `countls` targets to share the loop with `round`

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_bits.h:172]: nit: no `{}` for this inner `if` statement.

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_bits.h:83]: we should refactor this to be available for the whole `FXBits` class, and include the sign mask.

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_bits.h:173]: isn't this just take the complement of the bits? Does the operator `~` work for fixed point types?

Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h:173]: > 4.1.6.1.2
> ...
> It is not allowed to apply the complement
operator (~) to a fixed-point operand. 
> ...

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_bits.h:178]: Since you already clear the sign bit for negative numbers, maybe just `& TOTAL_MASK` would be more efficient (aka optimized to no-op) for signed types.

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h:83]: So, make `VALUE_MASK` a constant in `FXBits` just like how we already have `FRACTION_MASK`, etc.

Then, include an additional `TOTAL_MASK` that includes the sign bit as well.

Is that right?

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h:178]: So rename `get_value_bits` to `get_bits` and
inside `get_bits` we `& TOTAL_MASK` instead of `& VALUE_MASK`.

The idea is...`TOTAL_MASK` is actually `StorageType::MAX` in most cases, and so clang can easily see its useless and optimize away.

Whereas it can't as easily get optimized away if we use `VALUE_MASK` because we discard the sign bit.

Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:178]: yes, that's correct.

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:83]: yes, that's correct.

File: libc/src/__support/fixed_point/fx_bits.h
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:173]: That's an odd one, something maybe we should revise when we try to add fixed point types to C standard.  I guess they were worried about integer promotion rules.  Anyway, you can try:
```
  x = FXBits(static_cast<StorageType>(~value)).get_val();
```
to see if it works properly.

Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h:173]: After thinking about it. You're right. This is equivalent to bitwise complement. 

I also found this function in `fx_bits.h`.

https://github.com/llvm/llvm-project/blob/f71e13eb354d6a02710ceec254e8c399c60ffdbf/libc/src/__support/fixed_point/fx_bits.h#L143-L150

Assuming that `static_cast` and `bit_cast` are free, this is one less operation than the current way I do it.

So `static_cast` should be free, as it seems its only used as a type hint here.
I think `bit_cast` is free by definition.

File: libc/src/stdfix/CMakeLists.txt
Comment (duncpro) [reply libc/src/stdfix/CMakeLists.txt:5]: This is fixed in 97377a4.

https://github.com/llvm/llvm-project/blob/97377a4ee6a92cd1a46c1d57d7c7ee75d18f8790/libc/test/src/stdfix/CMakeLists.txt#L45-L75

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h]: This is fixed in [cc87a9a](https://github.com/llvm/llvm-project/commit/cc87a9ac973131db1047a22ebfb2b0db2d2f1926).

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L165-L168

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h]: This is fixed in [cc87a9a](https://github.com/llvm/llvm-project/commit/cc87a9ac973131db1047a22ebfb2b0db2d2f1926).

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L165-L168

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h]: This is fixed in [cc87a9a](https://github.com/llvm/llvm-project/commit/cc87a9ac973131db1047a22ebfb2b0db2d2f1926).

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L54-L57

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L77-L79

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L170-L171

File: libc/src/__support/fixed_point/fx_bits.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_bits.h]: This is fixed in [cc87a9a](https://github.com/llvm/llvm-project/commit/cc87a9ac973131db1047a22ebfb2b0db2d2f1926).

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L54-L57

File: libc/src/stdfix/CMakeLists.txt
Comment (duncpro) [reply libc/src/stdfix/CMakeLists.txt]: This is fixed in cc87a9a.

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/stdfix/CMakeLists.txt#L33-L57

File: libc/src/__support/fixed_point/fx_rep.h
Comment (PiJoules) [inline libc/src/__support/fixed_point/fx_rep.h:48]: nit: Perhaps to reduce copying `VALUE_LEN` N times in each specialization, you could have a free function

```
template <typename T>
constexpr int GetValueLen() { return T::INTEGRAL_LEN + T::FRACTIONAL_LEN; }
```

File: libc/test/src/stdfix/CountlsTest.h
Comment (PiJoules) [inline libc/test/src/stdfix/CountlsTest.h:28]: I'm probably being picky here, but could you also add some test cases for boring trivial values like -0.5, -0.25, 10, 42, etc.

**Review by PiJoules — COMMENTED** [review]
Just a few nits/requests then I'd be happy to give an LGTM after:

1. For each of the `LIST_COUNTLS_TESTS`, could you also add one for the saturation equivalent of each test? Saturation shouldn't affect the result but it would be nice to assert this is the case.
2. Clang also has a mode `-fpadding-on-unsigned-fixed-point` which makes the number of fractional bits for each unsigned fixed point type the same as the number of fractional bits for each corresponding signed type, effectively adding a byte of padding for unsigned types. This is an optional feature that allows us to omit a shift when converting between equivalent signed + unsigned types. Would it also be possible to have a run through each of the tests that enables this flag? I imagine in cmake we could do something like `check_cxx_compiler_flag(-fpadding-on-unsigned-fixed-point)` to first check if we can build a variant of tests with this flag.

File: libc/src/__support/fixed_point/fx_rep.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_rep.h]: Resolved by [1d0488a](https://github.com/llvm/llvm-project/pull/114318/commits/1d0488a6fa8735b1bb8dc08a561eaf40e64ce581)

File: libc/src/__support/fixed_point/fx_rep.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_rep.h:300]: needs `LIBC_INLINE`

File: libc/test/src/stdfix/CountlsTest.h
Comment (duncpro) [reply libc/test/src/stdfix/CountlsTest.h:30]: See [e73b2a2](https://github.com/llvm/llvm-project/pull/114318/commits/e73b2a20c1f04f8c0c184279cc90ecf963781ef0).

Is this sufficient?

File: libc/src/__support/fixed_point/fx_rep.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_rep.h]: Resolved by [81148c2](https://github.com/llvm/llvm-project/pull/114318/commits/81148c241c9da38a6b9d586fe1addb486b5de04f)

**Comment by duncpro** [comment]
> For each of the `LIST_COUNTLS_TESTS`, could you also add one for the saturation equivalent of each test? Saturation shouldn't affect the result but it would be nice to assert this is the case.

AFAIK `countlsfx` is not defined for `_Sat` (saturating) types.

Screenshot 2024-11-05 at 5 37 24 PM

#
However, saturating arithmetic can be requested for non-`_Sat` fixed-point types via a `#pragma`.

Screenshot 2024-11-05 at 5 41 09 PM


But it appears `clang` does not support this pragma yet ([feature-tracking issue](https://github.com/llvm/llvm-project/issues/43241)).

Screenshot 2024-11-05 at 5 38 11 PM

#

@PiJoules Can you give guidance?

**Comment by PiJoules** [comment]
> > For each of the `LIST_COUNTLS_TESTS`, could you also add one for the saturation equivalent of each test? Saturation shouldn't affect the result but it would be nice to assert this is the case.
> 
> AFAIK `countlsfx` is not defined for `_Sat` (saturating) types.
> 
> Screenshot 2024-11-05 at 5 37 24 PM
> However, saturating arithmetic can be requested for non-`_Sat` fixed-point types via a `#pragma`.
> 
> Screenshot 2024-11-05 at 5 41 09 PM
> But it appears `clang` does not support this pragma yet ([feature-tracking issue](https://github.com/llvm/llvm-project/issues/43241)).
> 
> Screenshot 2024-11-05 at 5 38 11 PM
> @PiJoules Can you give guidance?

Oh what I mean was having some tests that pass saturating types to each of the functions taking non-saturating equivalent types, so something like:

```
_Sat short _Fract f = 0.5;
int i = countlshr(f);  // Should return the same value even if `f` was `short _Fract`
``` 

I wanted this just to assert that the result is correct independent of the saturation, but in hindsight I think that might not be appropriate here since that would be more along the lines of testing `_Sat`/non-`_Sat` conversion rather than the actual function implementation, so disregard my suggestion of adding the `_Sat` tests.

File: libc/test/src/stdfix/CountlsTest.h
Comment (PiJoules) [reply libc/test/src/stdfix/CountlsTest.h:30]: Yup that works for me. Thanks!

**Comment by duncpro** [comment]
@PiJoules Testing `-fpadding-on-unsigned-fixed-point` was a good idea.

The tests fail when this flag is enabled, and after investigating, I don't see any obvious reason.

For instance, this assertion fails for `unsigned short fract`...
```cpp
 EXPECT_EQ(0, func(max)); // EXPECT_EQ(0, 1)
```

I'll continue to investigate. In the meantime if anyone else has any ideas as to why this happens please let me know.

File: libc/src/__support/fixed_point/fx_rep.h
Comment (lntue) [inline libc/src/__support/fixed_point/fx_rep.h:248]: also remove this in favor of `get_value_len()`?

File: libc/src/stdfix/CMakeLists.txt
Comment (lntue) [inline libc/src/stdfix/CMakeLists.txt:58]: Why do we need these extra entry points built with this flag?  Are they supposed to be on or off by default?
If it's only for testing with and without the `-Xclang=-fpadding-on-unsigned-fixed-point`, then using `FLAGS` in the main entry points is more apt.

**Review by PiJoules — APPROVED** [review]
LGTM on my end. Thanks! Let's get @lntue's thumbs up before landing.

File: libc/src/stdfix/CMakeLists.txt
Comment (duncpro) [reply libc/src/stdfix/CMakeLists.txt:58]: > Why do we need these extra entry points built with this flag? Are they supposed to be on or off by default? If it's only for testing with and without the `-Xclang=-fpadding-on-unsigned-fixed-point`, then using `FLAGS` in the main entry points is more apt.

Yes, the idea is to run all countlsfx tests twice. Once with `-Xclang=-fpadding-on-unsigned-fixed-point` and once without. I'm not sure what you mean by "using `FLAGS` in the main entry point". 

Are there any other places in the codebase where we run the same test suite twice but differ the flags? If so, I can mirror what is done there.

File: libc/src/__support/fixed_point/fx_rep.h
Comment (duncpro) [reply libc/src/__support/fixed_point/fx_rep.h]: Yes, I must've missed this one. Thanks.

Resolved by [04db3f7](https://github.com/llvm/llvm-project/pull/114318/commits/04db3f7b880ebc904d0829e36b652f19886db6d3)

File: libc/cmake/modules/CheckCompilerFeatures.cmake
Comment (michaelrj-google) [inline libc/cmake/modules/CheckCompilerFeatures.cmake:148]: nit: looks like you added an extra space here

File: libc/cmake/modules/compiler_features/check_padding_on_unsigned_fixed_point.cpp
Comment (michaelrj-google) [inline libc/cmake/modules/compiler_features/check_padding_on_unsigned_fixed_point.cpp:1]: is this file supposed to have more? Maybe an `ifdef` to check the proper values are being defined? Notably, if you don't pass `-ffixed-point` to clang this will compile successfully, since `stdfix-macros.h` checks that before defining anything else.

File: libc/cmake/modules/CheckCompilerFeatures.cmake
Comment (michaelrj-google) [inline libc/cmake/modules/CheckCompilerFeatures.cmake:67]: I think you also need to add `-ffixed-point` here.

**Review by michaelrj-google — COMMENTED** [review]
I think there are some issues with your compile flag checking. Also @lntue do you have any comments on setting this up to use flags?

**Pull request closed without merging**
